[MINI-5826] Update Demo app Settings to support multiple emails for contacts

GondiTilak opened the pull request:
# Description
Update our Demo app to show that we can support multiple emails for contacts.
Settings>Contacts
Now contact list shows the additional email ids.
Now on create new contact will add one additional email id to allEmailList.
User can add the new email id to the allEmailList when creating new contact or editing contact.
User can delete the additional email ids in edit screen.

## Links
[MINI-5826](https://jira.rakuten-it.com/jira/browse/MINI-5826)
[Demo Video on Box](https://rak.box.com/s/jds69cb6xkvsi7eb3t69qbt81pwcgbud)

# Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `fastlane ci` without errors

rleojoseph reviewed (commented):
Nice, Looks good to me :+1:
I have a few comments though,
1. Please update the placeholder text as "Email Address (Optional)" as mentioned in the ticket.
2. Do not add the new optional email address automatically when creating a new contact.

Could you please update the video with the same contacts retrieving in the JS Sample miniapp ?

GondiTilak commented:
> Nice, Looks good to me 👍 I have a few comments though,
> 
> 1. Please update the placeholder text as "Email Address (Optional)" as mentioned in the ticket. - Updated in the latest commit.
> 2. Do not add the new optional email address automatically when creating a new contact. - Removed the fakeEmailList() and no additional emails will be created by default.
> 
> Could you please update the video with the same contacts retrieving in the JS Sample miniapp ? - Its updated with the latest video at [Updated video link](https://rak.box.com/s/jds69cb6xkvsi7eb3t69qbt81pwcgbud)

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -120,6 +127,28 @@ extension MiniAppSettingsContactsView {
+                    .keyboardType(.emailAddress)
+                ForEach(allEmailList.indices, id: \.self) { index in
+                    HStack {
+                        TextField("Email Address (Optional) \(index + 1)", text: $allEmailList[index])
```
```suggestion
                        TextField("E-mail Address (Optional) \(index + 1)", text: $allEmailList[index])
```

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -169,12 +198,34 @@ extension MiniAppSettingsContactsView {
                 errorMessage += "Email id cannot be empty.\n"
```
```suggestion
                errorMessage += "E-mail id cannot be empty.\n"
```

GondiTilak replied:
This is now update.

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -169,12 +198,34 @@ extension MiniAppSettingsContactsView {
+            for (idx, emailId) in allEmails.enumerated() {
+                if !emailId.isValueEmpty() {
+                    if !emailId.isValidEmail() {
+                        errorMessage += "Email \(idx + 1) is invalid.\n"
```
```suggestion
                        errorMessage += "E-mail \(idx + 1) is invalid.\n"
```

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -169,12 +198,34 @@ extension MiniAppSettingsContactsView {
+                        errorMessage += "Email \(idx + 1) is invalid.\n"
+                    }
+                } else {
+                    errorMessage += "Email \(idx + 1) cannot be empty.\n"
```
```suggestion
                    errorMessage += "E-mail \(idx + 1) cannot be empty.\n"
```

GondiTilak replied:
I think doing this will include primary email in the list of additional emails. Which will have the redundant data in the list. And may create the confusion when using the additional emails list. This will also change the behavior of the allEmailList as per my understanding. and this list will always have one value.

GondiTilak replied:
Thank you :) I have update this in new commit.

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -169,12 +198,34 @@ extension MiniAppSettingsContactsView {
+                        errorMessage += "Email \(idx + 1) is invalid.\n"
+                    }
+                } else {
+                    errorMessage += "Email \(idx + 1) cannot be empty.\n"
```
```suggestion
                    errorMessage += "E-mail \(idx + 1) cannot be empty.\n"
```

GondiTilak replied:
I think doing this will include primary email in the list of additional emails. Which will have the redundant data in the list. And may create the confusion when using the additional emails list.

Climbatize commented on Tests/Unit/Helpers.swift:
```diff
@@ -413,7 +413,7 @@ class MockMessageInterface: MiniAppMessageDelegate {
             name: MiniAppStore.randomFakeName(),
             email: MiniAppStore.fakeMail(with: MiniAppStore.randomFakeName()),
-            allEmailList: MiniAppStore.fakeMailList(with: MiniAppStore.randomFakeName())
+            allEmailList: ["test@test.com"]
```
Why not testing with a list of more than 1 element? :)

GondiTilak replied:
Thank you :) I have update this in new commit.

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -15,7 +15,8 @@ struct MiniAppSettingsContactsView: View {
                         contactId: contact.id,
-                        email: contact.email ?? ""
+                        email: contact.email ?? "",
+                        allEmails: contact.allEmailList ?? [""]
```
```suggestion
                        allEmails: contact.allEmailList ?? [contact.email ?? ""]
```

GondiTilak replied:
I think doing this will include primary email in the list of additional emails. Which will have the redundant data in the list. And may create the confusion when using the additional emails list. This will also change the behavior of the allEmailList as per my understanding. and this list will always have one value.

Climbatize replied:
I get you, but having "" as primary e-mail and [""] as a list will lead to the same thing, right?

GondiTilak replied:
Yes but, I think for UI or any other case we can completely ignore the "", and this is easy to handle when we add the empty string checks also in case of JS-Sample.

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -15,7 +15,8 @@ struct MiniAppSettingsContactsView: View {
                         contactId: contact.id,
-                        email: contact.email ?? ""
+                        email: contact.email ?? "",
+                        allEmails: contact.allEmailList ?? [""]
```
```suggestion
                        allEmails: contact.allEmailList ?? [contact.email ?? ""]
```

GondiTilak replied:
I think doing this will include primary email in the list of additional emails. Which will have the redundant data in the list. And may create the confusion when using the additional emails list. This will also change the behavior of the allEmailList as per my understanding. and this list will always have one value.

Climbatize replied:
I suggest you go with an empty list then `[]`

GondiTilak replied:
Yes this worked nicely now I am using [] instead of [""]. Thank you. :)

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -98,6 +101,9 @@ extension MiniAppSettingsContactsView {
                     Text("Contact ID: \(contactId)")
                         .lineLimit(1)
                     Text("E-Mail address: \(email)")
+                    if !allEmails.isEmpty && allEmails.first != "" {
```
```suggestion
                    if allEmails.first ?? "" != "" {
```

GondiTilak replied:
Thank you :) I have update this in new commit.

Climbatize reviewed (approved):
LGTM. Look at my suggestions

Climbatize commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsContactsView.swift:
```diff
@@ -65,6 +67,20 @@ struct MiniAppSettingsContactsView: View {
+
+    func removeEmptyEmailStrings(for contact: MAContact) -> MAContact {
+        var newContact = contact
+        if newContact.allEmailList != nil {
+            if newContact.allEmailList?.first ?? "" == "" {
+                newContact.allEmailList = nil
+            } else {
+                newContact.allEmailList = newContact.allEmailList?.filter { $0 != "" }
+            }
+            return newContact
+        } else {
+            return newContact
+        }
```
Sorry I did not notice this before, this should work too
```suggestion
        if newContact.allEmailList?.first ?? "" == "" {
            newContact.allEmailList = nil
        } else {
            newContact.allEmailList = newContact.allEmailList?.filter { $0 != "" }
        }
        return newContact
```

GondiTilak replied:
Thanks, I did find a flaw in this and fixed it in new commit. ;)

rleojoseph reviewed (approved):
Looks good to me :+1:
